### PR TITLE
WIP: Add support for git push on every commit

### PR DESCRIPTION
**Pull request by grisuthedragon**
Since I am using my own git server instead of nextcloud to keep my files in sync, I often forgot to push the changes to my notes' folder. For this reason, I created this patch, which allows performing a `git push` after each commit. In this way, my remote git repository is kept up to date when using QOwnNotes. 

The patch adds an option to the notes folder in order to enable/disable the git push functionallity. This requires the database version of the notesFolder table to be updated to `42`.

### CODE REVIEW

**Comment by pbek** [comment]
As much I would like this to include, I think it still doesn't address the fundamental issue mentioned in https://github.com/pbek/QOwnNotes/issues/2076#issuecomment-803638320 or https://github.com/pbek/QOwnNotes/issues/744#issuecomment-336156850... The part with conflict resolution. I can already see issues like "QOwnNotes isn't pushing my commit, pls fix". 🙈 

Do you have a suggestion?

File: src/dialogs/settingsdialog.ui [hunk 1]
```diff
@@ -4756,7 +4766,7 @@ Just test yourself if you get sync conflicts and set a higher value if so.</stri
                 <string notr="true">&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
 &lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
 p, li { white-space: pre-wrap; }
-&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:'Noto Sans'; font-size:11pt; font-weight:400; font-style:normal;&quot;&gt;
```
Comment (pbek) [inline src/dialogs/settingsdialog.ui:4759]: Can you please not change the default font. 😁  I know, QtCreator tends to do this...

Comment (grisuthedragon) [reply src/dialogs/settingsdialog.ui:4759]: Sorry I tried to remove everything, which was not part of adding the checkbox, but it seems that I overlooked something.

Comment (pbek) [reply src/dialogs/settingsdialog.ui:4759]: One of the quirks of QtCreator...

**Comment by pbek** [comment]
Another issue, especially for new users of QOwnNotes and git, will be how to set a git remote...

**Comment by grisuthedragon** [comment]
For the beginning, I would say, they have to do this on the CLI. If the repository is clone, the remote is set any way.

**Comment by pbek** [comment]
> So the option I could think of would be to build a pull/push around the commit and make some options like "On conflict use mine" "On conflict use theirs" to obtain a default handling of the conflict situation. 

That would work... Another option then.

> For the beginning, I would say, they have to do this on the CLI. If the repository is clone, the remote is set any way.

Then we need a guide to explain the steps...
For example a `git.md` in https://github.com/pbek/QOwnNotes/tree/develop/webpage/src/getting-started...

**Comment by grisuthedragon** [comment]
I found another problem, which is causes trouble in case of the pulls. The detection if a file was changed on disk does not detect the stuff going on when pulling and thus changes are lost. 

So another idea, which may lead to an easier to handle solution. (Regarding where I was coming from, that I work on the notes, close the laptop and then recognized that the stuff was not pushed). 
 1. Add a Menu Entry or a Toolbar Button for "Git Pull", which raises an error message in case of a merge conflict. 
 2. Do the push on every commit, raising an error message, that a pull is required before. 
 
Does this sound reasonable?

**Comment by pbek** [comment]
Seems like there is a reason that tools like https://github.com/muesli/gitomatic/ exist. 😅 
I don't think anyone would like a manual pull.... Maybe just force-push (and possibly destroying anything on the remote repository)?

**Comment by grisuthedragon** [comment]
I tried a bit how to get a reasonable git workflow integrated, which tries to merge if possible, otherwise throws an error message. In a normal one-person workflow this seems to be reasonable. The only problem I have left now, is that after doing the pull/push, all currently open notes have to be reloaded in case the have been changed by the merge. @pbek  Can you help me implementing this?

**Comment by pbek** [comment]
> all currently open notes have to be reloaded in case the have been changed by the merge.

Doesn't that already work?
The note folder should reload automatically if files are changed if you do not actively block that FileWatcher and DirectoryWatcher event in the source code.

File: src/utils/git.cpp [hunk 2]
```diff
@@ -55,7 +55,7 @@ void Utils::Git::commitCurrentNoteFolder() {
     if (!executeGitCommand(git, QStringList{"init"}, process) ||
         !executeGitCommand(git, QStringList{"config", "commit.gpgsign", "false"}, process) ||
         !executeGitCommand(git, QStringList{"add", "-A"}, process) ||
-        !executeGitCommand(git, QStringList{"commit", "-m", "QOwnNotes commit"}, process)) {
+        !executeGitCommand(git, QStringList{"commit", "-m", "QOwnNotes commit"}, process, false, true)) {
```
Comment (pbek) [inline src/utils/git.cpp:58]: Showing an error dialog here will cause the user to be spammed with error messages every X seconds. And if the user isn't in front of a computer there maybe will be 1000s of dialogs 😅

File: src/utils/git.cpp [hunk 3]
```diff
@@ -79,7 +79,25 @@ void Utils::Git::pushCurrentNoteFolder() {
     auto* process = new QProcess();
     process->setWorkingDirectory(NoteFolder::currentLocalPath());
 
+    if (!executeGitCommand(git, QStringList{"pull", "--no-commit", "--no-ff"}, process)) {
```
Comment (pbek) [inline src/utils/git.cpp:82]: What if the error is "git command not found"?

File: src/utils/git.cpp [hunk 4]
```diff
@@ -79,7 +79,25 @@ void Utils::Git::pushCurrentNoteFolder() {
...
+                QObject::tr("The pull and merge of the remote repository failed. Please check on the command line whether the remote is configured correctly"
+                    "or if a manual merge is required."),"git-pull-merge-failed.");
+        delete(process);
+        return;
+    } else {
+        executeGitCommand(git, QStringList{"commit", "-am", "Merge commit"}, process, false, true);
+    }
+
     if (!executeGitCommand(git, QStringList{"push", "-q"}, process)) {
+        Utils::Gui::warning(
```
Comment (pbek) [inline src/utils/git.cpp:97]: Again, this will spam the user with error dialogs...

File: src/utils/git.cpp (hunk 2 above)
Comment (grisuthedragon) [reply src/utils/git.cpp:58]: That one does not show up with error messages, the second boolean parameter means, that the return value is ignored.

File: src/utils/git.cpp (hunk 4 above)
Comment (grisuthedragon) [reply src/utils/git.cpp:97]: Moving this to the status bar?

**Comment by grisuthedragon** [comment]
> > all currently open notes have to be reloaded in case the have been changed by the merge.
> 
> Doesn't that already work? The note folder should reload automatically if files are changed if you do not actively block that FileWatcher and DirectoryWatcher event in the source code.

Unfortunately not. I tried it in the following way. I changed a file at the beginning in the git remote and in the end using Qownnotes. Thus, the merge will be successful. In order to see the change, I have to close the note and open it again.

File: src/utils/git.cpp (hunk 3 above)
Comment (grisuthedragon) [reply src/utils/git.cpp:82]: Since the commit is done before, there would be an error message earlier.

File: src/utils/git.cpp (hunk 4 above)
Comment (pbek) [reply src/utils/git.cpp:97]: It will be hard to do that from a library (`git.cpp`).

**Comment by pbek** [comment]
> Unfortunately not. I tried it in the following way. I changed a file at the beginning in the git remote and in the end using Qownnotes. Thus, the merge will be successful. In order to see the change, I have to close the note and open it again.

...or reload the note folder (see *Note* menu)
Or in code like in `MainWindow::on_action_Reload_note_folder_triggered()`, but maybe not reload the current note, that might be annoying while you are typing into the note...

**Comment by grisuthedragon** [comment]
> > > all currently open notes have to be reloaded in case the have been changed by the merge.
> > 
> > 
> > Doesn't that already work? The note folder should reload automatically if files are changed if you do not actively block that FileWatcher and DirectoryWatcher event in the source code.
> 
> Unfortunately not. I tried it in the following way. I changed a file at the beginning in the git remote and in the end using Qownnotes. Thus, the merge will be successful. In order to see the change, I have to close the note and open it again.

It seems that I was too fast during testing. Now it works.

**Comment by grisuthedragon** [comment]
I moved the error and status message to the status bar in the main window. Now it seems to work, at least for my purpose.

File: src/utils/git.cpp (hunk 2 above)
Comment (grisuthedragon) [reply src/utils/git.cpp:59]: The errors are now shown in the status bar. So there will be no flood of dialogs.

Comment (grisuthedragon) [reply src/utils/git.cpp:59]: > ~Showing an error dialog here will cause the user to be spammed with error messages every X seconds. And if the user isn't in front of a computer there maybe will be 1000s of dialogs~
> 
> Ah, new parameter

Yes, to distinguish between "allowed/accepted" errors and the ones which are fatal.

File: src/utils/git.cpp (hunk 4 above)
Comment (grisuthedragon) [reply src/utils/git.cpp]: I pass a pointer from the mainwindow in, when the commit or pull is called.

File: src/utils/git.cpp [hunk 5]
```diff
@@ -23,6 +23,7 @@
 
 #include "gui.h"
 #include "misc.h"
+#include "mainwindow.h"
```
Comment (pbek) [inline src/utils/git.cpp:26]: MainWindow in a utility library... 🙈

**Comment by pbek** [comment]
When I activate the "git push checkbox" (with the default settings of 10sec interval) QOwnNotes takes about 19sec to start (instead of 0.1sec). When I start typing (or even if I don't do anything at all) the UI is frozen for a long time, then responds for a sec or so just to be unresponsive again...

I had this issue when an external git push wasn't correctly setup (remote git repository on gitlab.com) and also after it was setup and no error was shown in QOwnNotes. Most of the time I had to kill the QOwnNotes process. I'm not even able to interact with the application after the startup anymore.

Everything still works fine when I don't use the "put push" feature.

PS: It should not be possible to enable the "git push" checkbox when "git support" isn't enabled in the settings dialog.

**Comment by pbek** [comment]
> the UI is frozen for a long time,

Because `process->waitForFinished()` is running the process in the main thread to wait for the result.

**Comment by pbek** [comment]
...a QTimer::singleShot around the whole remote git pull/push operation (and being very diligent what gets accessed in that time) should help.

and because the remote git pull/push can take ages there also needs to be some kind of locking so that not multiple git pull/push operations can be active at the same time (and cause havoc on git and DOS the local computer and the server)

File: src/utils/git.cpp [hunk 6]
```diff
@@ -50,12 +56,59 @@ void Utils::Git::commitCurrentNoteFolder() {
     if (!executeGitCommand(git, QStringList{"init"}, process) ||
         !executeGitCommand(git, QStringList{"config", "commit.gpgsign", "false"}, process) ||
         !executeGitCommand(git, QStringList{"add", "-A"}, process) ||
-        !executeGitCommand(git, QStringList{"commit", "-m", "QOwnNotes commit"}, process)) {
+        !executeGitCommand(git, QStringList{"commit", "-m", "QOwnNotes commit"}, process, false, true)) {
+        if ( mw != nullptr ) {
```
Comment (pbek) [inline src/utils/git.cpp:60]: Can you please use the same coding style for brackets and spaces like in the rest of the project. 😉

**Comment by pbek** [comment]
> a QTimer::singleShot around the whole remote git pull/push operation (and being very diligent what gets accessed in that time) should help.

`QTimer::singleShot` still blocks the UI thread. 😬  seems like QThread needs to be used to run the commit and push methods.

**Comment by pbek** [comment]
I will close this until there is more information.

**Pull request closed without merging**